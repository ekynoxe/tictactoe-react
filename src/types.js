export default {
    singlePlayer: 1,
    twoPlayersLocal: 2
};
