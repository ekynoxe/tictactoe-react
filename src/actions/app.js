import actionTypes from './actionTypes';

export default {
    start: function start(time) {
        return {
            type: actionTypes.APP.STARTED,
            time: time
        };
    }
};
