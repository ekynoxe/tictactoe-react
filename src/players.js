export default {
    o: 'o',
    x: 'x'
};
