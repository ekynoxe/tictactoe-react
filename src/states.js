export default {
    inplay: 0,
    draw: 1,
    win: 2
};
