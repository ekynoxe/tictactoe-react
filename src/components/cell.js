import React from 'react';
import { connect } from 'react-redux';
import actions from '../actions';

class BaseCell extends React.Component {
    select() {
        this.props.dispatch(actions.selectCell(this.props.index));
    }

    render() {
        let mark;

        if (this.props.mark) {
            mark = this.props.mark;

        } else {
            mark = <button onClick={ this.select.bind(this) }>&nbsp;</button>;
        }

        return(<div className='cell'><span className='mark'>{ mark }</span></div>);
    }
}

export const Cell = connect()(BaseCell);
