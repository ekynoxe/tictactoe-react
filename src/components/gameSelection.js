import React from 'react';
import { connect } from 'react-redux';
import gameTypes from '../types';
import actions from '../actions';

export class GameSelection extends React.Component {
    selectGame(type) {
        this.props.dispatch(actions.selectGame(type));
    }

    render() {
        return <div className='buttons'>
            <button key='btn_single_player' className='button' onClick={ this.selectGame.bind(this, gameTypes.singlePlayer) }>Single player</button>
            <button key='btn_two_players_local' className='button' onClick={ this.selectGame.bind(this, gameTypes.twoPlayersLocal) }>Two players in this window</button>
        </div>;
    }
}

export default connect()(GameSelection);
