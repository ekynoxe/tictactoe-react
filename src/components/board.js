import React from 'react';
import { Cell } from './cell';

export class Board extends React.Component {
    render() {
        let rows = [];
        let cells;

        this.props.board.forEach( (m, i) => {
            if (!(i % 3)) {
                cells = [];
            }

            cells.push(<Cell key={ `c_${i}` } mark={ m } index={ i } />);

            if (!((i + 1) % 3)) {
                let row = Math.floor(i / 3);
                rows.push(<div key={ `r_${row}` } className='row'>{ cells }</div>);
            }
        });

        return(<div className='board'>{ rows }</div>);
    }
}
