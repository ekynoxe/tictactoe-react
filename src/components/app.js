import React from 'react';
import { connect } from 'react-redux';
import { Board } from './board';
import actions from '../actions';
import players from '../players';
import states from '../states';

class BaseApp extends React.Component {
    reset() {
        this.props.dispatch(actions.reset());
    }

    setPlayer(player) {
        this.props.dispatch(actions.setPlayer(player));
    }

    componentWillUpdate(nextProps) {
        // On first load or reset, no player will be selected, so we randomly
        //  select a player to start the game here.
        if (!nextProps.currentPlayer) {
            this.setPlayer(Math.round(Math.random()) === 0 ? players.o : players.x);
        }
    }

    render() {
        // Waiting for application to load the local store
        if (!this.props.rehydrated) {
            return <div>Loading...</div>;
        }

        let notice;

        if (states.inplay !== this.props.gameState) {
            notice = 'Game over (declare winner here...)';

        } else if (this.props.currentPlayer) {
            notice = `Current player: ${ this.props.currentPlayer.toUpperCase() }`;
        }

        return(<div>
            <Board board={ this.props.board } />
            <p className='notice'>{ notice }</p>
            <button onClick={ this.reset.bind(this) }>Restart</button>
        </div>);
    }
}

function mapStateToProps(state) {
    return {
        board: state.board,
        currentPlayer: state.currentPlayer,
        gameState: state.gameState
    };
}

export const App = connect(mapStateToProps)(BaseApp);
