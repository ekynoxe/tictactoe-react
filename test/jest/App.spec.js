/* global jest, test, expect */
import React from 'react';
import renderer from 'react-test-renderer';
import { App } from '../../src/components/app';
import { inPlayBoard } from '../boards';
import players from '../../src/players';
import states from '../../src/states';
import types from '../../src/types';

jest.mock('../../src/components/board', () => 'Board');
jest.mock('../../src/components/cell', () => 'Cell');

test('App matches snapshot', () => {
    const component = renderer.create(
        <App rehydrated={ true } board={ inPlayBoard } currentPlayer={ players.x } gameState={ states.inplay } gameType={ types.singlePlayer } winner={ null } />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
