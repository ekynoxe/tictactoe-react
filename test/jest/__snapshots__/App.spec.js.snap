// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot 1`] = `
<div>
  <Board
    board={
      Array [
        "x",
        "o",
        "x",
        "o",
        "o",
        "x",
        null,
        null,
        null,
      ]
    }
  />
  <p
    className="notice"
  >
    Current player: X
  </p>
  <button
    onClick={[Function]}
  >
    Restart
  </button>
</div>
`;
